korim/src/commonMain/kotlin/com/soywiz/korim/bitmap/BitmapSlice.kt

```diff
@@ -250,29 +294,33 @@ fun <T : ISizeInt> BmpCoordsWithT<T>.named(name: String?): BmpCoordsWithInstance
  *                     the bounds rectangle. Width and height defines the size of the virtual frame.
  */
 abstract class BmpSlice(
-    val bmpBase: Bitmap,
+    bmpBase: Bitmap,
     val bounds: RectangleInt,
     override val name: String? = null,
-    val rotated: Boolean = false,
-    val virtFrame: RectangleInt? = null
+    imageOrientation: ImageOrientation = ImageOrientation.ORIGINAL,
+    final override val virtFrame: RectangleInt? = null,
+    parentCoords: BitmapCoords? = null,
+    bmpCoords: BmpCoordsWithT<ISizeInt>? = null
 ) : Extra, BitmapCoords {
-    override val base get() = bmpBase
-    open val bmp: Bitmap = bmpBase
-    val bmpWidth = bmpBase.width
-    val bmpHeight = bmpBase.height
 
-    private val tl = Point(left.toFloat() / bmpBase.width.toFloat(), top.toFloat() / bmpBase.height.toFloat())
-    private val br = Point(right.toFloat() / bmpBase.width.toFloat(), bottom.toFloat() / bmpBase.height.toFloat())
-    private val tr = Point(br.x, tl.y)
-    private val bl = Point(tl.x, br.y)
+    @Deprecated("Use imageOrientation instead of rotation")
+    constructor(
+        bmpBase: Bitmap,
+        bounds: RectangleInt,
+        name: String? = null,
+        rotated: Boolean = false,
+        virtFrame: RectangleInt? = null
+    ): this(bmpBase, bounds, name, if (rotated) ImageOrientation.ROTATE_90 else ImageOrientation.ORIGINAL, virtFrame)
 
-    private val points = arrayOf(tl, tr, br, bl)
-    private val offset = if (rotated) 1 else 0
+    var bmpBase: Bitmap = bmpBase
+        private set
+    var imageOrientation: ImageOrientation = imageOrientation
+        private set
 
-    private val p0 = points.getCyclic(offset + 0)
-    private val p1 = points.getCyclic(offset + 1)
-    private val p2 = points.getCyclic(offset + 2)
-    private val p3 = points.getCyclic(offset + 3)
```

Comment:
Can't we use these points somehow or an array of points and represent the `ImageOrientation` as a reorder of those points?

Like flippedX, flippedY:

https://github.com/korlibs/korge-next/blob/44300cc63f05d20ff25bf94a077bbb47b29ee704/korim/src/commonMain/kotlin/com/soywiz/korim/bitmap/BitmapSlice.kt#L215-L225

The `init { }` method in this PR looks pretty complicated to me and doing a lot of things with lots of edge cases and I'm a bit afraid of introducing it and having to maintain it later.

I'm also wondering if we really need to put this complexity here. In fact those methods are already doing the required transformations.
What if instead of returning `BmpSlice` in the atlas, we return values from those methods, and we apply the required transformations there?

BmpCoordsWithT was introduced precisely for that, to keep BmpSlice simple, and maybe we could even deprecate the rotated property and ignore it. What would you think? Is something preventing from returning that in the atlas? Since we are bumping version to 3.0.0 and it is a major release, we can afford to change return types

In fact in KorGE views we alreaady changed:
https://github.com/korlibs/korge-next/blob/44300cc63f05d20ff25bf94a077bbb47b29ee704/korim/src/commonMain/kotlin/com/soywiz/korim/bitmap/BitmapSlice.kt#L84

To use BaseBmpSlice:
https://github.com/korlibs/korge-next/blob/44300cc63f05d20ff25bf94a077bbb47b29ee704/korge/src/commonMain/kotlin/com/soywiz/korge/view/Image.kt#L46

That is the type returned from those methods.